packages/tpetra/core/src/Tpetra_Import_def.hpp

```diff
@@ -297,96 +318,88 @@ namespace Tpetra {
       } catch (Teuchos::Exceptions::InvalidParameter&) {}
     }
     debug_ = debug;
+    ArrayView<const GO> sourceGIDs = source->getNodeElementList ();
+    ArrayView<const GO> targetGIDs = target->getNodeElementList ();
+    const size_type numSrcGids = sourceGIDs.size ();
+    const size_type numTgtGids = targetGIDs.size ();
+    const size_type numGids = std::min (numSrcGids, numTgtGids);
+    const LO LINVALID = Teuchos::OrdinalTraits<LO>::invalid ();
 
-    if (debug_ && ! out_.is_null ()) {
-      std::ostringstream os;
-      const int myRank = source->getComm ()->getRank ();
-      os << myRank << ": constructExpert " << std::endl;
-      *out_ << os.str ();
-    }
     ImportData_ = rcp (new data_type (source, target, out_, plist));
-    ImportData_->numSameIDs_ = numSameGids;
 
-    Array<LO>& permuteToLIDs = ImportData_->permuteToLIDs_;
-    Array<LO>& permuteFromLIDs = ImportData_->permuteFromLIDs_;
-    Array<LO>& remoteLIDs = ImportData_->remoteLIDs_;
-    const LO LINVALID = Teuchos::OrdinalTraits<LO>::invalid ();
-    const LO numTgtLids = as<LO> (numTgtGids);
+    Array<GO>  tRemoteGIDs;
 
-    if(!useRemotePIDGID) {
-      remoteGIDs.clear();
-      remoteLIDs.clear();
-    }
+    setupSamePermuteRemote(tRemoteGIDs);
 
-    for (LO tgtLid = numSameGids; tgtLid < numTgtLids; ++tgtLid) {
-      const GO curTargetGid = targetGIDs[tgtLid];
-      // getLocalElement() returns LINVALID if the GID isn't in the source Map.
-      const LO srcLid = source->getLocalElement (curTargetGid);
-      if (srcLid != LINVALID) {
-        permuteToLIDs.push_back (tgtLid);
-        permuteFromLIDs.push_back (srcLid);
-      } else {
-        if(!useRemotePIDGID) {
-          remoteGIDs.push_back (curTargetGid);
-          remoteLIDs.push_back (tgtLid);
-        }
+    Array<LO>& tRemoteLIDs = ImportData_->remoteLIDs_;
+    Teuchos::Array<int> tRemotePIDs(userRemotePIDs);
+
+    TPETRA_ABUSE_WARNING(
+                         getNumRemoteIDs() > 0 && ! source->isDistributed(),
+                         std::runtime_error,
+                         "::constructExpert(): Target has remote LIDs but Source is not "
+                         "distributed globally." << std::endl
+                         << "Importing to a submap of the target map.");
+    TEUCHOS_TEST_FOR_EXCEPTION
+      ( !(tRemotePIDs.size() == tRemoteGIDs.size() &&
+          tRemoteGIDs.size() == tRemoteLIDs.size()),
+        std::runtime_error,
+        "Import::Import createExpert version: Size miss match on userRemotePIDs, remoteGIDs and remoteLIDs Array's to sort3. This will produce produce an error, aborting ");
```

Comment:
The last sentence is a bit silly.  Why not print the sizes?  (I won't insist on this for the PR, though.)